File: golang/q800/q830.go
package q800

func largeGroupPositions(s string) [][]int {
	r := [][]int{}
	l := 0
	begin := 0
	for i := 1; i < len(s); i++ {
		if s[i] == s[i-1] {
			if l == 0 {
				begin = i - 1
			}
			l++
		} else {
			if l >= 2 {
				r = append(r, []int{begin, i - 1})
			}
			l = 0
		}
	}
	if l >= 2 {
		r = append(r, []int{begin, len(s) - 1})
	}
	return r
}

File: hs/src/Q800/Q830.hs
module Q800.Q830 (
    largeGroupPositions
) where

import Data.List(group)

largeGroupPositions :: String -> [[Int]]
largeGroupPositions = let 
        plana  (tl, r) l = (tl+l, (tl,l):r)
    in 
    map (\(begin,l) -> [begin,begin+l-1]) . filter (\(_,l) -> l >=3 ) . snd . foldl plana (0,[]) . map length . group